File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.py
dy = [0, -1, 0, 1]
dx = [-1, 0, 1, 0]

from collections import deque


def is_corner(pre_y, pre_x, y, x, ny, nx):
    if pre_y == -1 or pre_x == -1:
        return False

    if (pre_y == y == ny) or (pre_x == x == nx):
        return False
    else:
        return True


def bfs(cost, board, n):
    que = deque()
    cost[0][0] = [100] * 4
    que.append((-1, -1, 0, 0, 0))
    while que:
        pre_y, pre_x, y, x, now_cost = que.popleft()
        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]

            if 0 <= ny < n and 0 <= nx < n:
                pass
            else:
                continue

            if pre_y == ny and pre_x == nx:
                continue

            if board[ny][nx] == 1:
                continue

            sum_cost = now_cost
            if is_corner(pre_y, pre_x, y, x, ny, nx):
                sum_cost += 600
            else:
                sum_cost += 100

            if sum_cost >= cost[ny][nx][i]:
                continue

            cost[ny][nx][i] = sum_cost
            que.append((y, x, ny, nx, sum_cost))

    return min(cost[-1][-1])


def solution(board):
    n = len(board)
    cost = [[[float('inf')] * 4 for _ in range(n)] for _ in range(n)]
    return bfs(cost, board, n)


# print(solution([[0, 0, 0], [0, 0, 0], [0, 0, 0]]))

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java
import java.util.stream.*;
import java.util.*;

class Solution {
    public int[] dy = new int[] {0, -1, 0, 1};
    public int[] dx = new int[] {-1, 0, 1, 0};
    public int[][][] cost;
    
    public boolean isCorner(int preY, int preX, int y, int x, int ny, int nx){
        if(preY == -1 && preX == -1){
            return false;
        }
        
        if ((preY == y && y == ny) || (preX == x && x == nx)){
            return false;
        }
        
        return true;
    }
    
    
    public int bfs(int[][] board){
        LinkedList<int[]> que = new LinkedList<>();
        
        que.add(new int[]{-1,-1, 0, 0, 0});
        
        while(!que.isEmpty()){
            int[] ints = que.pollFirst();
            
            for(int i = 0; i < 4; i++){
                int ny = ints[2] + dy[i];
                int nx = ints[3] + dx[i];
                
                if (ny < 0 || ny >= cost.length || nx < 0 || nx >= cost.length) continue;
                
                if(ints[0] == ny && ints[1] == nx) continue;
                
                if (board[ny][nx] == 1) continue;
                    
                int sum = ints[4] + 100;
                if (isCorner(ints[0], ints[1], ints[2], ints[3], ny, nx)){
                    sum += 500;
                }
                
                if (sum >= cost[ny][nx][i]) continue;
                
                cost[ny][nx][i] = sum;
                que.addLast(new int[]{ints[2], ints[3], ny, nx, sum});
            }
        }
        
        return (int)Arrays.stream(cost[cost.length - 1][cost.length - 1]).min().orElse(0);
    }
    
    
    public int solution(int[][] board) {
        cost = new int[board.length][board.length][4];
        
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board.length; j++){
                for(int k = 0; k < 4; k ++){
                    cost[i][j][k] = 2000000000;
                }
            }
        }
        
        for(int i = 0; i < 4; i++){
            cost[0][0][i] = 0;
        }
        
        return bfs(board);
    }
}